flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/ml/SqlVectorSearchTableFunction.java

```diff
@@ -0,0 +1,239 @@
+import java.util.Optional;
+
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.toLogicalType;
+
+/**
+ * {@link SqlVectorSearchTableFunction} implements an operator for search.
+ *
+ * <p>It allows four parameters:
+ *
+ * <ol>
+ *   <li>a table
+ *   <li>a descriptor to provide a column name from the input table
+ *   <li>a query column from the left table
+ *   <li>a literal value for top k
+ * </ol>
+ */
+public class SqlVectorSearchTableFunction extends SqlFunction implements SqlTableFunction {
+
+    private static final String PARAM_SEARCH_TABLE = "SEARCH_TABLE";
+    private static final String PARAM_COLUMN_TO_SEARCH = "COLUMN_TO_SEARCH";
+    private static final String PARAM_COLUMN_TO_QUERY = "COLUMN_TO_QUERY";
+    private static final String PARAM_TOP_K = "TOP_K";
+
+    private static final String OUTPUT_SCORE = "score";
+
+    public SqlVectorSearchTableFunction() {
+        super(
+                "VECTOR_SEARCH",
+                SqlKind.OTHER_FUNCTION,
+                ReturnTypes.CURSOR,
+                null,
+                new OperandMetadataImpl(),
+                SqlFunctionCategory.SYSTEM);
+    }
+
+    @Override
+    public SqlReturnTypeInference getRowTypeInference() {
+        return new SqlReturnTypeInference() {
+            @Override
+            public @Nullable RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+                final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+                final RelDataType inputRowType = opBinding.getOperandType(0);
+
+                return typeFactory
+                        .builder()
+                        .kind(inputRowType.getStructKind())
+                        .addAll(inputRowType.getFieldList())
+                        .addAll(
+                                SqlValidatorUtils.makeOutputUnique(
+                                        inputRowType.getFieldList(),
+                                        Collections.singletonList(
+                                                new RelDataTypeFieldImpl(
+                                                        OUTPUT_SCORE,
+                                                        0,
+                                                        typeFactory.createSqlType(
+                                                                SqlTypeName.DOUBLE)))))
+                        .build();
+            }
+        };
+    }
+
+    @Override
+    public boolean argumentMustBeScalar(int ordinal) {
+        return ordinal != 0;
+    }
+
+    private static class OperandMetadataImpl implements SqlOperandMetadata {
+
+        private static final List<String> PARAMETERS =
+                Collections.unmodifiableList(
+                        Arrays.asList(
+                                PARAM_SEARCH_TABLE,
+                                PARAM_COLUMN_TO_SEARCH,
+                                PARAM_COLUMN_TO_QUERY,
+                                PARAM_TOP_K));
```

Comment:
What is TOP_K? It would be useful to add in the documentation describing the parameters with this change - including the default. Is this the maximum number of results that come back? If so I wonder if we should add paging, to be able to handle a large number of results.